modules/nuxeo-apidoc-core/src/main/java/org/nuxeo/apidoc/snapshot/SnapshotListener.java

```diff
@@ -0,0 +1,156 @@
+/*
+ * (C) Copyright 2020 Nuxeo (http://nuxeo.com/) and others.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ * Contributors:
+ *     Anahide Tchertchian
+ */
+package org.nuxeo.apidoc.snapshot;
+
+import static org.nuxeo.runtime.ComponentEvent.COMPONENT_STARTED;
+import static org.nuxeo.runtime.ComponentEvent.EXTENSION_REGISTERED;
+
+import java.io.Serializable;
+import java.util.HashMap;
+import java.util.Map;
+import java.util.concurrent.atomic.AtomicInteger;
+
+import org.nuxeo.apidoc.api.ExtensionInfo;
+import org.nuxeo.apidoc.api.ExtensionPointInfo;
+import org.nuxeo.apidoc.introspection.RuntimeSnapshot;
+import org.nuxeo.runtime.ComponentEvent;
+import org.nuxeo.runtime.ComponentListener;
+import org.nuxeo.runtime.api.Framework;
+import org.nuxeo.runtime.model.Component;
+import org.nuxeo.runtime.model.ComponentInstance;
+import org.nuxeo.runtime.model.ComponentManager;
+import org.nuxeo.runtime.model.ComponentManager.Listener;
+import org.nuxeo.runtime.model.ComponentStartOrders;
+import org.nuxeo.runtime.model.Extension;
+
+/**
+ * Listens for component events and stores information about it at runtime startup.
+ * <p>
+ * This allows to detect resolution/registration orders on component contributions to be contributed to the live
+ * snapshot {@link RuntimeSnapshot} information.
+ * <p>
+ * Events will stop being handled after the framework has started.
+ *
+ * @since 20.0.0
+ */
+public class SnapshotListener implements ComponentListener, Listener {
+
+    // records framework startup to avoid handling later hot-reload events
+    protected boolean listening = true;
+
+    protected int startCounter = -1;
+
+    // component name -> start order
+    protected final Map<String, Integer> starts = new HashMap<>();
+
+    // component name -> declared start order
+    protected final Map<String, Integer> declaredStarts = new HashMap<>();
+
+    // component extension point -> integer (to handle multiple contributions to the same extension point)
+    protected final Map<String, AtomicInteger> compLocalCounters = new HashMap<>();
+
+    // extension point id -> extension registration counter
+    protected final Map<String, AtomicInteger> xpCounters = new HashMap<>();
+
+    // extension id -> extension registration counter
+    protected final Map<String, Integer> registrations = new HashMap<>();
```

Comment:
What about changing Integer to Long as we're converting them in the end?

Reply:
I actually can't remember why i previously used integers on existing similar counters, and kept it as is so that it's similar, maybe because it helps with increment operations? (?)

Reply:
Are you talking about `AtomicInteger` and `AtomicLong`?

Reply:
Yes, about existing logics on ExtensionInfoImpl constructor here: https://github.com/nuxeo/nuxeo-explorer/blob/v20.0.0-RC1/modules/nuxeo-apidoc-core/src/main/java/org/nuxeo/apidoc/introspection/ServerInfo.java#L504
I'm ok moving this to AtomicLong if there's no actual drawback.

Reply:
done